Fixes #22181 - Update katello-change-hostname to work with custom certs

johnpmitsch opened the pull request:
For plugin updates, please indicate which repos this should be built into:

* [x] Nightly
* [ ] 1.16
* [ ] 1.15
* [ ] 1.14

See Foreman's [plugin maintainer documentation](http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Release-strategies) for more information.

---

ekohl commented on katello/katello/hostname-change.rb:
```diff
@@ -161,6 +161,16 @@ def get_default_proxy_id
+    def custom_cert_options_check(scenario_answers)
+      if scenario_answers["certs"]["server_cert"] &&
+         scenario_answers["certs"]["server_key"] &&
+         scenario_answers["certs"]["server_cert_req"] &&
```
The trailing ```&&``` is odd here. Technically it'll work but I'd rather get rid of it.

johnpmitsch replied:
+1 had something after it that I removed and I forgot to delete the "&&"

johnpmitsch commented:
From the commit message:
katello-change-hostname fails with custom certs currently. Currently
there are no arguments to pass new certs that match the new hostname
to the installer. This commit adds those options and some checks    
to see if the user is using custom certs. If they are using custom
certs, katello-certs-check will run before any changes are made.
On proxies w/ content, a certs tar must be made with the new
custom certs. This is passed into the existing --certs-tar option.

johnpmitsch commented:
To review:
Please test the following scenarios plus any edge cases you can think of.

- [ ] change hostname on an install with custom certs is successful
- [ ] change hostname on a proxy w/ content with custom certs is successful
- [ ] change hostname on a normal katello install is successful
- [ ] change hostname on a normal proxy w/ content install is successful

I'm happy to walk through any or all of the scenarios with a reviewer

johnpmitsch commented:
1. set up a blank centos7 box
1. Follow instructions on setting up custom certs (I will send you an article).
2. Once the custom certs are created, you will need to install the katello repositories and yum install katello as specified in [our docs](https://theforeman.org/plugins/katello/3.5/installation/index.html). I used 3.5 to test, you can also use nightly, but I wouldn't test on any older version as the code will be too far out of sync.
3. When you get to the installer step, pass in the custom certs as arguments. There should be an example custom cert article that is roughly the same.
4. Generate new certs with a new hostname using the same 'ownca' tool mentioned in the article. You won't need to generate a new CA here, just using `./generate-crt.sh <HOSTNAME>`. We will now use them in the hostname change.
5. Run katello-change-hostname with the new certs. i.e.
`katello-change-hostname -u admin -p changeme -c /cert/changed.example.com.crt -k /cert/changed.example.com.key -r /cert/changed.example.com.crt.req -b /cert/cacert.crt  -y changed.example.com`
6. The hostname of the katello box should be changed and katello should function properly

johnpmitsch commented:
For a foreman proxy w/ content (this is fine to do after changing the hostname on the original katello box) :
1. spin up another centos7 box, we will install the foreman-proxy here
2. on the katello box, generate new certs that match the hostname of the new box. (use ./generate-crt.sh again)
3. on the katello box, run capsule-certs-generate, but pass in the new certs generated for the proxy: i.e.
`foreman-proxy-certs-generate --server-ca-cert /certs/cacert.crt --server-cert /certs/customcertscapsule.john.example.com.crt --server-cert-req /certs/customcertscapsule.john.example.com.crt.req --server-key /certs/customcertscapsule.john.example.com.key --certs-tar /root/customcertscapsule.tar.gz --foreman-proxy-fqdn "customcertscapsule.john.example.com"`
3. Use the output to install the capsule as normal. This will give you a proxy installed with custom certs.
4. To run katello-change-hostname on the proxy, repeat steps 2 and 3 with a new hostname (can be any valid fqdn).
5. Ignore the output from foreman-proxy-certs-generate, just copy the certs tar that was generated to the capsule.
6. Run katello-change-hostname on the proxy with the new certs tar: i.e.
`./katello-change-hostname -u admin -p changeme -c /root/changedcapsule.tar.gz -y changedcapsule.example.com`

johnpmitsch commented:
installer fails w/o full path being passed in:
```
[ERROR 2018-01-10 20:44:13 verbose] Parameter certs-server-cert invalid: newname.cfoo.example.com.crt is not one of regexes matching /^(([a-zA-Z]:[\\\/])|([\\\/][\\\/][^\\\/]+[\\\/][^\\\/]+)|([\\\/][\\\/]\?[\\\/][^\\\/]+))/ or regexes matching /^\/([^\/\0]+\/*)+$/
[ERROR 2018-01-10 20:44:13 verbose] Parameter certs-server-key invalid: newname.cfoo.example.com.key is not one of regexes matching /^(([a-zA-Z]:[\\\/])|([\\\/][\\\/][^\\\/]+[\\\/][^\\\/]+)|([\\\/][\\\/]\?[\\\/][^\\\/]+))/ or regexes matching /^\/([^\/\0]+\/*)+$/
[ERROR 2018-01-10 20:44:13 verbose] Parameter certs-server-cert-req invalid: newname.cfoo.example.com.crt.req is not one of regexes matching /^(([a-zA-Z]:[\\\/])|([\\\/][\\\/][^\\\/]+[\\\/][^\\\/]+)|([\\\/][\\\/]\?[\\\/][^\\\/]+))/ or regexes matching /^\/([^\/\0]+\/*)+$/
```
We can check if the file is valid and expand the path for them, then use that to pass into the installer

cfouant commented:
@johnpmitsch ACK - these work great! thank you! @ekohl we need you to approve the requested changes made before this can merge.

ekohl commented on katello/katello/helper.rb:
```diff
@@ -36,6 +38,12 @@ def fail_with_message(message, opt_parser=nil)
     end
 
+    def check_if_file_exists(file)
+      unless File.exist?(file) && (Pathname.new file).absolute?
```
Nit: I prefer ```Pathname.new(file).absolute?```

ekohl commented on katello/katello/hostname-change.rb:
```diff
@@ -211,6 +233,22 @@ def run
+                          "you need to provide certificates with the new hostname. Please review the custom certificate arguments required with --help.")
+      end
+
+      if @using_custom_certs && !@foreman_proxy_content
```
The interaction here a bit hard to follow. The above check can fail if the options aren't provided but here you exclude if for proxies. Does that mean we should also exclude it above?

ekohl commented on katello/katello/hostname-change.rb:
```diff
@@ -161,6 +161,16 @@ def get_default_proxy_id
+      if scenario_answers["certs"]["server_cert"] &&
+         scenario_answers["certs"]["server_key"] &&
+         scenario_answers["certs"]["server_cert_req"]
+        @using_custom_certs = true
```
I don't like this side effect here. Maybe it should be ```custom_cert_options_check!```. It could also raise an exception if it's invalid.

ekohl commented on katello/katello/helper.rb:
```diff
@@ -36,6 +38,12 @@ def fail_with_message(message, opt_parser=nil)
       exit(false)
     end
 
+    def check_if_file_exists(file)
```
This has a side effect so should this be ```check_if_file_exists!``` for clarity?

ekohl commented on katello/katello/hostname-change.rb:
```diff
@@ -211,6 +233,22 @@ def run
+      if @using_custom_certs && !@foreman_proxy_content
+        STDOUT.puts "\nChecking custom certificates"
+        [@options[:custom_cert], @options[:custom_key], @options[:custom_cert_req]].each do |file|
+          check_if_file_exists(file)
```
Would it be nice if we checked all files and failed with one message that some files don't exist?

johnpmitsch commented:
@ekohl @cfouant thanks for the suggestions, I updated a few things to improve the usability.
The certs check will now accept relative paths in addition to absolute paths. It also now tells all the files that are missing (rather than one at a time) here are some examples:

File name is correct, but I'm in the wrong directory to use absolute paths:
```
[root@customcertsupstream katello]# ./katello-change-hostname -u admin -p changeme -c /certs/changedupstream.example.com.crt -k /certs/changedupstream.example.com.key -r changedupstream.example.com.crt.req changed.example.com

Checking custom certificates
Error: File  /root/foreman-packaging/katello/katello/changedupstream.example.com.crt.req does not exist! Please check the file path and try again.
```

Same situation as above, but now two files:
 ```
[root@customcertsupstream katello]# ./katello-change-hostname -u admin -p changeme -c /certs/changedupstream.example.com.crt -k changedupstream.example.com.key -r changedupstream.example.com.crt.req changed.example.com

Checking custom certificates
Error: Files  /root/foreman-packaging/katello/katello/changedupstream.example.com.key, /root/foreman-packaging/katello/katello/changedupstream.example.com.crt.req do not exist! Please check the file path and try again.
```

Changed to the correct directory and now using relative paths (successful run):
```
[root@customcertsupstream certs]# ~/foreman-packaging/katello/katello/katello-change-hostname -u admin -p changeme -c changedupstream.example.com.crt -k changedupstream.example.com.key -r changedupstream.example.com.crt.req changed.example.com

Checking custom certificates

Checking hostname validity
```

ekohl commented on katello/katello/hostname-change.rb:
```diff
@@ -209,6 +231,24 @@ def setup_opt_parser
+        end
+      end
+
+      if using_custom_certs?(scenario_answers) && !@foreman_proxy_content
```
Can you elaborate why the certs only need to be checked in non-proxy context?

johnpmitsch replied:
Yes, the foreman proxy hostname change requires a new certs-tarball whether the user is using custom certs or not. This tarball is passed into k-c-h as a single argument and custom certs are included in this tarball. 

Unfortunately there isn't a good way to check the certs in the tarball right now, we should make katello-certs-check work for proxies and check these sorts of things in the future.

cfouant commented:
ACK
